scripts/gethimport.py

```diff
@@ -0,0 +1,619 @@
+
+    if db_already_existed:
+        return MainnetChain(leveldb)
+
+    logger.info(f'Trinity database did not already exist, initializing it now')
+    chain = MainnetChain.from_genesis_header(leveldb, MAINNET_GENESIS_HEADER)
+
+    logger.warining('The new db contains the genesis header but not the genesis state.')
+    logger.warining('Attempts to full sync will fail.')
+
+    return chain
+
+
+def import_headers(gethdb, chain):
+    headerdb = chain.headerdb
+
+    logger.warning('Some features are not yet implemented:')
+    logger.warning('- This only supports importing the mainnet chain')
+    logger.warning('- This script will not verify that geth is using the mainnet chain')
+
+    canonical_head = headerdb.get_canonical_head()
+    logger.info(f'starting import from trinity\'s canonical head: {canonical_head}')
+
+    # fail fast if geth disagrees with trinity's canonical head
+    geth_header = gethdb.block_header(canonical_head.block_number, canonical_head.hash)
+    assert geth_header.hash == canonical_head.hash
+
+    geth_last_block_hash = gethdb.last_block_hash
+    geth_last_block_num = gethdb.block_num_for_hash(geth_last_block_hash)
+
+    final_block_to_sync = geth_last_block_num
+    if args.syncuntil:
+        final_block_to_sync = min(args.syncuntil, final_block_to_sync)
+
+    for i in range(canonical_head.block_number, final_block_to_sync + 1):
+        header_hash = gethdb.header_hash_for_block_number(i)
+        header = gethdb.block_header(i, header_hash)
+        headerdb.persist_header(header)
+
+        if i % 1000 == 0:
+            logger.debug(f'current canonical header: {headerdb.get_canonical_head()}')
+
+    canonical_head = headerdb.get_canonical_head()
+    if not args.syncuntil:
+        # similar checks should be run if we added sync until!
+        # some final checks, these should never fail
+        assert canonical_head.hash == geth_last_block_hash
+        assert canonical_head.block_number == geth_last_block_num
+
+    logger.info('finished importing headers + bodies')
+
+
+def sweep_state(gethdb: GethDatabase, trinitydb: LevelDB):
+    """
+    Imports state, but by indiscriminately copying over everything which might be part of
+    the state trie. This copies more data than necessary, but is likely to be much faster
+    than iterating all state.
+    """
+    logger.debug('sweep_state: bulk-importing state entries')
+
+    iterator = gethdb.db.iterator(
+        start=b'\x00'*32,
+        stop=b'\xff'*32,
+        include_start=True,
+        include_stop=True,
+    )
+
+    imported_entries = 0
+    skipped_keys = 0
+    bucket = b'\x00' * 2
+    for key, value in iterator:
+        if len(key) != 32:
+            skipped_keys += 1
+            continue
+        trinitydb[key] = value
+        imported_entries += 1
+
+        if key >= bucket:
+            logger.debug(f'imported: {bucket.hex()} skipped={skipped_keys}')
+            if bucket == b'\xff' * 2:
```

Comment:
This is a bug that I fixed on my server but not locally, I'll push a new commit soon!